File: backend/src/main/java/com/leavemarker/entity/Payment.java
package com.leavemarker.entity;

import com.leavemarker.enums.BillingCycle;
import com.leavemarker.enums.PaymentStatus;
import com.leavemarker.enums.PaymentType;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "payments", indexes = {
    @Index(name = "idx_payment_transaction_id", columnList = "transactionId"),
    @Index(name = "idx_payment_razorpay_order_id", columnList = "razorpayOrderId"),
    @Index(name = "idx_payment_razorpay_payment_id", columnList = "razorpayPaymentId"),
    @Index(name = "idx_payment_company", columnList = "company_id"),
    @Index(name = "idx_payment_status", columnList = "status")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Payment extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subscription_id", nullable = false)
    private Subscription subscription;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "company_id", nullable = false)
    private Company company;

    // Unique internal transaction ID
    @Column(nullable = false, unique = true, length = 100)
    private String transactionId;

    // Razorpay Order ID (created before payment)
    @Column(length = 100)
    private String razorpayOrderId;

    // Razorpay Payment ID (after successful payment)
    @Column(length = 100)
    private String razorpayPaymentId;

    // Razorpay Signature (for verification)
    @Column(length = 256)
    private String razorpaySignature;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PaymentType paymentType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BillingCycle billingCycle;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal amount;

    @Column(precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal taxAmount = BigDecimal.ZERO;

    @Column(precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal totalAmount;

    @Column(length = 10)
    @Builder.Default
    private String currency = "INR";

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private PaymentStatus status = PaymentStatus.PENDING;

    @Column(length = 100)
    private String paymentMethod;

    // Timestamps for tracking
    private LocalDateTime initiatedAt;

    private LocalDateTime paidAt;

    private LocalDateTime failedAt;

    private LocalDateTime refundedAt;

    // Period this payment covers
    @Column(nullable = false)
    private LocalDateTime periodStart;

    @Column(nullable = false)
    private LocalDateTime periodEnd;

    // Webhook tracking
    @Column(length = 50)
    private String webhookStatus;

    private LocalDateTime webhookReceivedAt;

    @Column(columnDefinition = "TEXT")
    private String webhookPayload;

    // Audit info
    @Column(length = 50)
    private String ipAddress;

    @Column(length = 500)
    private String userAgent;

    @Column(columnDefinition = "TEXT")
    private String metadata;

    @Column(length = 500)
    private String failureReason;

    @Column(length = 500)
    private String refundReason;

    // Idempotency key for retries
    @Column(length = 100, unique = true)
    private String idempotencyKey;

    // Retry tracking
    @Column(nullable = false)
    @Builder.Default
    private Integer retryCount = 0;

    private LocalDateTime lastRetryAt;
}

File: backend-go/internal/models/payment.go
package models

import (
	"time"

	"github.com/leavemarker/backend-go/internal/enums"
	"github.com/shopspring/decimal"
)

// Payment represents a payment entity
type Payment struct {
	BaseModel
	SubscriptionID     *uint               `gorm:"index" json:"subscriptionId"`
	PlanID             uint                `gorm:"not null;index" json:"planId"`
	CompanyID          uint                `gorm:"not null;index" json:"companyId"`
	TransactionID      string              `gorm:"size:100;uniqueIndex;not null" json:"transactionId"`
	RazorpayOrderID    *string             `gorm:"size:100" json:"razorpayOrderId"`
	RazorpayPaymentID  *string             `gorm:"size:100" json:"razorpayPaymentId"`
	RazorpaySignature  *string             `gorm:"size:256" json:"razorpaySignature"`
	PaymentType        enums.PaymentType   `gorm:"size:20;not null" json:"paymentType"`
	BillingCycle       enums.BillingCycle  `gorm:"size:20" json:"billingCycle"`
	Amount             decimal.Decimal     `gorm:"type:decimal(10,2);not null" json:"amount"`
	TaxAmount          decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"taxAmount"`
	DiscountAmount     decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"discountAmount"`
	TotalAmount        decimal.Decimal     `gorm:"type:decimal(10,2);not null" json:"totalAmount"`
	Currency           string              `gorm:"size:10;default:INR" json:"currency"`
	Status             enums.PaymentStatus `gorm:"size:20;default:PENDING" json:"status"`
	PaymentMethod      *string             `gorm:"size:100" json:"paymentMethod"`

	// Timestamps
	InitiatedAt *time.Time `json:"initiatedAt"`
	PaidAt      *time.Time `json:"paidAt"`
	FailedAt    *time.Time `json:"failedAt"`
	RefundedAt  *time.Time `json:"refundedAt"`

	// Period Coverage
	PeriodStart time.Time `json:"periodStart"`
	PeriodEnd   time.Time `json:"periodEnd"`

	// Webhook Tracking
	WebhookStatus     *string    `gorm:"size:50" json:"webhookStatus"`
	WebhookReceivedAt *time.Time `json:"webhookReceivedAt"`
	WebhookPayload    *string    `gorm:"type:text" json:"webhookPayload"`

	// Audit
	IPAddress     *string `gorm:"size:50" json:"ipAddress"`
	UserAgent     *string `gorm:"size:500" json:"userAgent"`
	Metadata      *string `gorm:"type:text" json:"metadata"`
	FailureReason *string `gorm:"size:500" json:"failureReason"`
	RefundReason  *string `gorm:"size:500" json:"refundReason"`

	// Retry
	IdempotencyKey *string    `gorm:"size:100;uniqueIndex" json:"idempotencyKey"`
	RetryCount     int        `gorm:"default:0" json:"retryCount"`
	LastRetryAt    *time.Time `json:"lastRetryAt"`

	// Relationships
	Subscription *Subscription `gorm:"foreignKey:SubscriptionID" json:"-"`
	Plan         Plan          `gorm:"foreignKey:PlanID" json:"-"`
	Company      Company       `gorm:"foreignKey:CompanyID" json:"-"`
}

func (Payment) TableName() string {
	return "payments"
}
